cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,233 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package upgrader
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/compute"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	"google.golang.org/api/option"
+)
+
+// Parameter key shared with external packages
+const (
+	ClientIDFlagKey = "client-id"
+	DefaultTimeout  = "90m"
+)
+
+const (
+	logPrefix = "[windows-upgrade]"
+
+	metadataKeyWindowsStartupScriptURL       = "windows-startup-script-url"
+	metadataKeyWindowsStartupScriptURLBackup = "windows-startup-script-url-backup"
+
+	versionWindows2008r2 = "windows-2008r2"
+	versionWindows2012r2 = "windows-2012r2"
+)
+
+var (
+	upgradeScriptName = map[string]string{versionWindows2008r2: "upgrade_script_2008r2_to_2012r2.ps1"}
+
+	expectedCurrentLicense = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2008-r2-dc"}
+	licenseToAdd           = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2012-r2-dc-in-place-upgrade"}
+)
+
+type derivedVars struct {
+	instanceProject string
+	instanceZone    string
+	instanceURI     string
+
+	osDiskURI        string
+	osDiskType       string
+	osDiskDeviceName string
+	osDiskAutoDelete bool
+
+	instanceName           string
+	machineImageBackupName string
+	osDiskSnapshotName     string
+	newOSDiskName          string
+	installMediaDiskName   string
+
+	originalWindowsStartupScriptURL *string
+}
+
+// InputParams contains input params for the upgrade.
+type InputParams struct {
+	ClientID               string
+	ProjectPtr             *string
+	Zone                   string
+	Instance               string
+	SkipMachineImageBackup bool
+	AutoRollback           bool
+	SourceOS               string
+	TargetOS               string
+	Timeout                string
+	ScratchBucketGcsPath   string
+	Oauth                  string
+	Ce                     string
+	GcsLogsDisabled        bool
+	CloudLogsDisabled      bool
+	StdoutLogsDisabled     bool
+	CurrentExecutablePath  string
+}
+
+type upgrader struct {
+	*InputParams
+	*derivedVars
+
+	ctx context.Context
+}
+
+type upgraderInterface interface {
+	getUpgrader() *upgrader
+	init() error
+	validateAndDeriveParams() error
+	printIntroHelpText() error
+	prepare() (*daisy.Workflow, error)
```

Comment:
'prepare' and 'init' sound very similar! Can you please write out what each of these mean? For example: pre-conditions, post-conditions, what the goal of each phase is.